src/box/txn_limbo.c

```diff
@@ -237,6 +237,8 @@ txn_limbo_on_remove(struct txn_limbo *limbo,
 	assert(limbo->size >= 0);
 	limbo->len--;
 	assert(limbo->len >= 0);
+	/* Wake up all fiber waiting to add a new limbo entry. */
+	fiber_cond_broadcast(&limbo->wait_cond);
```

Comment:
It is going to be quite expensive to do that on each popped limbo entry. Because it would wakeup all the limbo-waiting fibers, including the `txn_limbo_wait_complete()`, and those we might have thousands. They would all get a spurious wakeup. This cond should only be used by rare events, like when the limbo parameters change.

A not too intrusive alternative would be to invoke it only when `limbo->len == limbo->max_size` (before the decrement). It means the queue was full and became not full. A state change, right? We are not interested in the cases, when the queue was full and remains full somehow. Or was not full and remains not full. Still a lot of spurious wakeups, but now only when the queue max size is reached, which is supposed to be rare.

Another way would be in `txn_add_limbo_entry()` to subscribe on the first limbo txn's on_commit trigger and the last txn's on_rollback trigger. In those triggers you would wake your fiber, and not any other fiber. And you would get notified when anything leaves the limbo. OTOH, if we get a lot of waiters and they all subscribe on the same border txns, they would all be woken up. Thus again spurious wakeups.

Unless you could think of any better ways, I would then suggest the hack above about `len == max_size` before decrement.

Reply:
I guess the ideal robust solution would be to do it similarly to `journal_queue` (refactoring it to some `fiber_write_queue` abstraction), but I would prefer going for a simpler solution that you mentioned for now.

> I would then suggest the hack above about `len == max_size` before decrement.

Done.